primary/src/tests/helper_tests.rs

```diff
@@ -180,6 +199,218 @@ async fn test_process_certificates_batch_mode() {
+                            assert_eq!(found, 7, "Expected to have available payloads");
+                        } else {
+                            assert_eq!(found, 3, "Expected to have non available payloads");
+                        }
+                    }
+                }
+                msg => {
+                    panic!("Didn't expect message {:?}", msg);
+                }
+            }
+        }
+    } else {
+        panic!(
+            "Timed out while waiting for results. Did not receive all the expected payload availability."
+        );
+    }
+}
+
+#[tokio::test]
+#[traced_test]
+async fn test_process_payload_availability_when_failures() {
+    // GIVEN
+    // We initialise the test stores manually to allow us
+    // inject some wrongly serialised values to cause data store errors.
+    let rocksdb = rocks::open_cf(temp_dir(), None, &[CERTIFICATES_CF, PAYLOAD_CF])
+        .expect("Failed creating database");
+
+    let (certificate_map, payload_map) = reopen!(&rocksdb,
+        CERTIFICATES_CF;<CertificateDigest, Certificate<Ed25519PublicKey>>,
+        PAYLOAD_CF;<(BatchDigest, WorkerId), PayloadToken>);
+
+    let certificate_store: Store<CertificateDigest, Certificate<Ed25519PublicKey>> =
+        Store::new(certificate_map);
+    let payload_store: Store<(types::BatchDigest, WorkerId), PayloadToken> =
+        Store::new(payload_map);
+
+    let key = keys().pop().unwrap();
+    let (name, committee) = resolve_name_and_committee(11340);
+    let (tx_primaries, rx_primaries) = channel(10);
+
+    // AND a helper
+    Helper::spawn(
+        committee.clone(),
+        certificate_store.clone(),
+        payload_store.clone(),
+        rx_primaries,
+    );
+
+    // AND some mock certificates
+    let mut certificate_ids = Vec::new();
+    for _ in 0..10 {
+        let header = fixture_header_builder()
+            .with_payload_batch(fixture_batch_with_transactions(10), 0)
+            .build(|payload| key.sign(payload));
+
+        let certificate = certificate(&header);
+        let id = certificate.clone().digest();
+
+        // In order to test an error scenario that is coming from the data store,
+        // we are going to store for the provided certificate ids some unexpected
+        // payload in order to blow up the deserialisation.
+        let serialised_key = bincode::DefaultOptions::new()
+            .with_big_endian()
+            .with_fixint_encoding()
+            .serialize(&id.borrow())
+            .expect("Couldn't serialise key");
+
+        // Just serialise the "false" value
+        let dummy_value = bincode::serialize(false.borrow()).expect("Couldn't serialise value");
+
+        rocksdb
+            .put_cf(
+                &rocksdb
+                    .cf_handle(CERTIFICATES_CF)
+                    .expect("Couldn't find column family"),
+                serialised_key,
+                dummy_value,
+            )
+            .expect("Couldn't insert value");
```

Comment:
Didn't find a better to inject/trigger an error for our storage to test the error path on the code. 

I am going to explore offline the https://docs.rs/mockall/latest/mockall/ so can see how we can handle such stuff in a more pretty way (and mocking in tests in general), but not for now.